vm/devices/storage/disk_nvme/nvme_driver/src/queue_pair.rs

```diff
@@ -461,6 +553,56 @@ impl QueueHandler {
             }
         }
     }
+
+    /// Save queue data for servicing.
+    pub async fn save(&self) -> anyhow::Result<QueuePairSavedState> {
+        let mut pending_cmds: Vec<PendingCommandSavedState> = Vec::new();
+        for cmd in &self.commands {
+            let mut command: [u8; 64] = [0; 64];
+            command.copy_from_slice(cmd.1.command.as_bytes());
+            let command = PendingCommandSavedState {
+                command,
+                cid: cmd.0 as u16,
+            };
+            pending_cmds.push(command);
+        }
+        // The data is collected from both QueuePair and QueueHandler.
+        Ok(QueuePairSavedState {
+            max_cids: self.max_cids,
+            sq_state: self.sq.save(),
+            cq_state: self.cq.save(),
+            pending_cmds,
+            cpu: 0,       // Will be updated by the caller.
+            msix: 0,      // Will be updated by the caller.
+            sq_addr: 0,   // Will be updated by the caller.
+            cq_addr: 0,   // Will be updated by the caller.
+            base_va: 0,   // Will be updated by the caller.
+            mem_len: 0,   // Will be updated by the caller.
+            pfns: vec![], // Will be updated by the caller.
+        })
+    }
+
+    /// Restore queue data after servicing.
+    pub fn restore(&mut self, saved_state: &QueuePairSavedState) -> anyhow::Result<()> {
+        self.max_cids = saved_state.max_cids;
+
+        // Restore pending commands.
+        let mut pending: Vec<(usize, PendingCommand)> = Vec::new();
+        for cmd in &saved_state.pending_cmds {
+            let (send, mut _recv) = mesh::oneshot::<nvme_spec::Completion>();
+            let pending_command = PendingCommand {
+                command: FromBytes::read_from_prefix(cmd.command.as_bytes()).unwrap(),
+                respond: send,
+            };
+            pending.push((cmd.cid as usize, pending_command));
```

Comment:
It seems like pending commands need to save the fact that they allocated memory, somehow.

Specifically, any PRP list allocation needs to be saved/restored. Is this done somewhere?

We also need to re-lock any guest memory/re-reserve any double buffer memory that was in use.

And of course, all this needs to be released after the pending command completes.

Reply:
If PRP points to the bounce buffer (vtl2 buffer) then its GPN should not change and it still points to the same physical page which is preserved during servicing.
Same should be true if PRP points to a region in vtl0 memory as we don't touch vtl0 during servicing.

However, maybe I missed the part on where we allocate PRP list itself, if there are multiple entries. If it is not a part of preserved memory then such change needs to be added.